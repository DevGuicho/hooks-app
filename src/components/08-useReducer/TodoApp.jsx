import React, { useEffect, useReducer } from "react";
import todoReducer from "./todoReducer";
import "./styles.css";
import TodoList from "./TodoList";
import TodoAdd from "./TodoAdd";

const init = () => {
  /* return [{ id: new Date().getTime(), desc: "Aprender react", done: false }]; */
  return JSON.parse(localStorage.getItem("todos")) || [];
};

const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleAddTodo = (desc) =>
    dispatch({
      type: "add",
      payload: {
        id: new Date().getTime(),
        desc,
        done: false,
      },
    });
  const handleDelete = (id) => {
    dispatch({
      type: "delete",
      payload: id,
    });
  };

  const handleToggle = (id) => {
    dispatch({
      type: "toggle",
      payload: id,
    });
  };

  return (
    <div className="container">
      <h1>Todo App ({todos.length}) </h1>
      <hr />
      <div className="row">
        <TodoList
          todos={todos}
          handleDelete={handleDelete}
          handleToggle={handleToggle}
        />
        <TodoAdd handleAddTodo={handleAddTodo} />
      </div>
    </div>
  );
};

export default TodoApp;
