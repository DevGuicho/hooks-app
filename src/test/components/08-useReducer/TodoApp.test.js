import { act } from "@testing-library/react";
import { mount, shallow } from "enzyme";
import TodoApp from "../../../components/08-useReducer/TodoApp";
import { demoTodos } from "../../fixtures/demoTodos";

describe("pruebas en <TodoApp />", () => {
  const wrapper = shallow(<TodoApp />);

  Storage.prototype.setItem = jest.fn();

  test("debe mostrar el componente correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de agregar un TODO", () => {
    const wrapper = mount(<TodoApp />);
    act(() => {
      wrapper.find("TodoAdd").prop("handleAddTodo")(demoTodos[0].desc);
      wrapper.find("TodoAdd").prop("handleAddTodo")(demoTodos[1].desc);
    });
    expect(wrapper.find("h1").text().trim()).toBe("Todo App (2)");
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });
});
