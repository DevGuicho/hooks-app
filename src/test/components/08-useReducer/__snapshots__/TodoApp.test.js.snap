// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <TodoApp /> debe mostrar el componente correctamente 1`] = `
<div
  className="container"
>
  <h1>
    Todo App (
    0
    ) 
  </h1>
  <hr />
  <div
    className="row"
  >
    <TodoList
      handleDelete={[Function]}
      handleToggle={[Function]}
      todos={Array []}
    />
    <TodoAdd
      handleAddTodo={[Function]}
    />
  </div>
</div>
`;
